xdem/coreg/affine.py

```diff
@@ -1160,102 +2214,150 @@ def _fit_rst_pts(
+            area_or_point=area_or_point,
+            z_name=z_name,
+            params_random=params_random,
+            params_fit_or_bin=params_fit_or_bin,
+            max_iterations=self._meta["inputs"]["iterative"]["max_iterations"],
+            tolerance=self._meta["inputs"]["iterative"]["tolerance"],
+            method=self._meta["inputs"]["specific"]["icp_method"],
+            picky=self._meta["inputs"]["specific"]["icp_picky"],
+            only_translation=self._meta["inputs"]["affine"]["only_translation"],
+            standardize=self._meta["inputs"]["affine"]["standardize"],
+        )
 
-        point_elev["E"] -= centroid[0]
-        point_elev["N"] -= centroid[1]
+        # Write output to class
+        # (Mypy does not pass with normal dict, requires "OutAffineDict" here for some reason...)
+        output_affine = OutAffineDict(
+            centroid=centroid,
+            matrix=matrix,
+            shift_x=matrix[0, 3],
+            shift_y=matrix[1, 3],
+            shift_z=matrix[2, 3],
+        )
+        self._meta["outputs"]["affine"] = output_affine
+        self._meta["outputs"]["random"] = {"subsample_final": subsample_final}
 
-        points["point"] = point_elev[["E", "N", z_name, "nx", "ny", "nz"]].values
 
-        for key in points:
-            points[key] = points[key][~np.any(np.isnan(points[key]), axis=1)].astype("float32")
-            points[key][:, 0] -= resolution[0] / 2
-            points[key][:, 1] -= resolution[1] / 2
+class CPD(AffineCoreg):
+    """
+    Coherent Point Drift coregistration, based on Myronenko and Song (2010), https://doi.org/10.1109/TPAMI.2010.46.
 
-        # Extract parameters and pass them to method
-        max_it = self._meta["inputs"]["iterative"]["max_iterations"]
-        tol = self._meta["inputs"]["iterative"]["tolerance"]
-        rej = self._meta["inputs"]["specific"]["rejection_scale"]
-        num_lv = self._meta["inputs"]["specific"]["num_levels"]
-        icp = cv2.ppf_match_3d_ICP(max_it, tol, rej, num_lv)
-        logging.info("Running ICP...")
-        try:
-            # Use points as reference
-            _, residual, matrix = icp.registerModelToScene(points["raster"], points["point"])
-        except cv2.error as exception:
-            if "(expected: 'n > 0'), where" not in str(exception):
-                raise exception
-
-            raise ValueError(
-                "Not enough valid points in input data."
-                f"'reference_dem' had {points['ref'].size} valid points."
-                f"'dem_to_be_aligned' had {points['tba'].size} valid points."
-            )
+    Estimate translations and rotations.
+
+    The estimated transform is stored in the `self.meta["outputs"]["affine"]` key "matrix", with rotation centered
+    on the coordinates in the key "centroid". The translation parameters are also stored individually in the
+    keys "shift_x", "shift_y" and "shift_z" (in georeferenced units for horizontal shifts, and unit of the
```

Comment:
Maybe the answer lies in the code, but if the rotation is not 0, this probably means translation of the centroid?